src/windows/osd.c

```diff
@@ -402,65 +402,111 @@ ssize_t sendmsg(SOCKET sd, struct msghdr *msg, int flags)
+{
+	PIP_ADAPTER_ADDRESSES ptr = head, tmp = NULL, prev = NULL;
+	bool sorted = false;
+	if (!head) {
+		return;
+	}
+	do {
+		sorted = false;
+		ptr = head;
+		while (ptr->Next) {
+			if (ptr->Ipv4Metric < ptr->Next->Ipv4Metric) {
+				if (ptr == head) {
+					tmp = ptr;
+					ptr = tmp->Next;
+					tmp->Next = ptr->Next;
+					ptr->Next = tmp;
+					head = ptr;
+					sorted = true;
+				}
+				else {
+					tmp = ptr;
+					ptr = tmp->Next;
+					tmp->Next = ptr->Next;
+					ptr->Next = tmp;
+					prev->Next = ptr;
+					sorted = true;
+				}
+			}
+			prev = ptr;
+			ptr = ptr->Next;
+		}
+	} while (sorted);
+}
+
 int getifaddrs(struct ifaddrs **ifap)
 {
-	DWORD i;
-	MIB_IPADDRTABLE _iptbl;
-	MIB_IPADDRTABLE *iptbl = &_iptbl;
-	ULONG ips = 1;
-	ULONG res = GetIpAddrTable(iptbl, &ips, 0);
+	DWORD res, i = 0;
+	DWORD size;
+	PIP_ADAPTER_ADDRESSES adapter_addresses, aa;
+	PIP_ADAPTER_UNICAST_ADDRESS ua;
 	int ret = -1;
 	struct ifaddrs *head = NULL;
 
-	assert(ifap);
-
-	if (res == ERROR_INSUFFICIENT_BUFFER) {
-		iptbl = malloc(ips);
-		if (!iptbl)
-			goto failed_no_mem;
-		res = GetIpAddrTable(iptbl, &ips, 0);
-		if (res != NO_ERROR)
-			goto failed_get_addr;
-	} else if (res != NO_ERROR) {
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX,
+				   NULL, NULL, &size);
+	if (res != ERROR_BUFFER_OVERFLOW) {
 		goto failed;
 	}
+	adapter_addresses = (PIP_ADAPTER_ADDRESSES)malloc(size);
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX,
+				   NULL, adapter_addresses, &size);
+	if (res != ERROR_SUCCESS) {
+		goto failed_cant_allocate;
+	}
 
-	for (i = 0; i < iptbl->dwNumEntries; i++) {
-		if (iptbl->table[i].dwAddr && iptbl->table[i].dwAddr != ntohl(INADDR_LOOPBACK)) {
+	sort_list_of_adapters(adapter_addresses);
+
+	for (aa = adapter_addresses; aa != NULL; aa = aa->Next) {
+		if (aa->OperStatus != 1) {
+			continue;
+		}
+		for (ua = aa->FirstUnicastAddress; ua != NULL; ua = ua->Next) {
+			SOCKADDR *pSockAddr = ua->Address.lpSockaddr;
+			if (pSockAddr->sa_family != AF_INET) {
+				continue;
+			}
 			struct ifaddrs *fa = calloc(sizeof(*fa), 1);
 			if (!fa)
 				goto failed_cant_allocate;
+
 			fa->ifa_flags = IFF_UP;
+			if (aa->IfType == IF_TYPE_SOFTWARE_LOOPBACK) {
+				fa->ifa_flags |= IFF_LOOPBACK;
+			}
 			fa->ifa_addr = (struct sockaddr *)&fa->in_addr;
 			fa->ifa_netmask = (struct sockaddr *)&fa->in_netmask;
 			fa->ifa_name = fa->ad_name;
 
-			fa->in_addr.sin_family = fa->in_netmask.sin_family = AF_INET;
-			fa->in_addr.sin_addr.s_addr = iptbl->table[i].dwAddr;
-			fa->in_netmask.sin_addr.s_addr = iptbl->table[i].dwMask;
-			/* on Windows there is no Unix-like interface names,
-			   so, let's generate fake names */
-			sprintf_s(fa->ad_name, sizeof(fa->ad_name), "eth%d", i);
-
+			fa->in_addr.sin_family = fa->in_netmask.sin_family = pSockAddr->sa_family;
+			ULONG subnet = 0;
+			PULONG mask = &subnet;
+			struct sockaddr_in *pInAddr = (struct sockaddr_in*)pSockAddr;
```

Comment:
Please keep all variable declarations at the top of the function.

Reply:
@shefty Done.